ml-agents/mlagents/trainers/trainer_controller.py

```diff
@@ -181,24 +182,35 @@ def write_to_tensorboard(self, global_step: int) -> None:
             else:
                 trainer.write_summary(global_step, delta_train_start)
 
+    def start_trainer(self, trainer: Trainer, env_manager: EnvManager) -> None:
+        self.trainers[trainer.brain_name] = trainer
+        self.logger.info(trainer)
+        if self.train_model:
+            trainer.write_tensorboard_text("Hyperparameters", trainer.parameters)
+        env_manager.set_policy(trainer.brain_name, trainer.policy)
+
     def start_learning(self, env_manager: EnvManager) -> None:
         self._create_model_path(self.model_path)
 
         tf.reset_default_graph()
 
-        for _, t in self.trainers.items():
-            self.logger.info(t)
-
         global_step = 0
+        last_brain_names: List[str] = []
 
-        if self.train_model:
-            for brain_name, trainer in self.trainers.items():
-                trainer.write_tensorboard_text("Hyperparameters", trainer.parameters)
         try:
-            for brain_name, trainer in self.trainers.items():
-                env_manager.set_policy(brain_name, trainer.policy)
             self._reset_env(env_manager)
             while self._not_done_training():
+                if last_brain_names != env_manager.external_brains.keys():
```

Comment:
I think this would be a little cleaner with sets instead of lists. Then you can do something like
``` python
external_brains = set(env_manager.external_brains.keys())
new_brains = external_brains - last_brain_names
for name in new_brains:
  ...
```